walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -43,56 +66,77 @@
 
         // remove available TimeRanges too small
         List<TimeRange> toRemove = new ArrayList<>();
-        
         for (int i = 0; i < availableRanges.size(); i++) {
             TimeRange available = availableRanges.get(i);            
             if (available.duration() < request.getDuration()) {
                 toRemove.add(available);
             }
         }
-
         availableRanges.removeAll(toRemove);
         
         return availableRanges;
+
   }
 
+
 /** 
 *returns set of TimeRanges formed by all events that have mandatory attendees from the request with no overlap
+* if mandatoryAttendees is true, we return the combinedRanges of mandatoryAttendees. else, return combinedRanges of optionalAttendees
 */
-  public Collection<TimeRange> combinedMandatoryRanges(Collection<Event> events, MeetingRequest request) {
+  public Collection<TimeRange> combinedRanges(Collection<Event> events, MeetingRequest request, boolean considerOptionalAttendees) {
         List<TimeRange> busyTimes = new ArrayList<>();
-        for (Event event : events) {
-            if (!Collections.disjoint(event.getAttendees(), request.getAttendees()) ) { //events with mandatory attendees only
-                busyTimes.add(event.getWhen());
+        if (considerOptionalAttendees) { 
+            for (Event event : events) {
+                if (!Collections.disjoint(event.getAttendees(), request.getAttendees()) || //events with mandatory attendees 
+                    !Collections.disjoint(event.getAttendees(), request.getOptionalAttendees())) { // or events with optional attendees
+                    busyTimes.add(event.getWhen());
+                }
+            }
+        } else { // combined events including optional attendees
+            for (Event event : events) {
+                if (!Collections.disjoint(event.getAttendees(), request.getAttendees()) ) { //events with mandatory attendees only
+                    busyTimes.add(event.getWhen());
+                }
             }
         }
+
         Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
 
         List<TimeRange> combinedBusyTimes = new ArrayList<>();
 
         int start = -1;
+        int latestEnd = 0;
+        int end = 0;
```

Comment:
If these changes fix a bug, there should be a unit test to verify the case that gets fixed by these changes

Reply:
Unit test added and patched hidden bug